### PR TITLE
Support configuring the library through environment variables

### PR DESCRIPTION
**Pull request by kattrali**
## Goal

Read `BUGSNAG_*` environment variables and configure the library automatically.

## Design

* For string-based properties, variables are accepted as-is. For example, the value of `BUGSNAG_API_KEY`, if set, is applied to `Configuration.APIKey`.
* For slice of string properties, commas are supported as a delimiter. Set `BUGSNAG_PROJECT_PACKAGES` to `main,example.com/repo` to mark code from the `main` and `example.com/repo` packages as in-project.
* For boolean properties, variables are checked against the string value `1`, and, if not empty string, are set to true if equal or false if not.

## Testing

Added new integration tests for each supported variable

### CODE REVIEW

File: features/configuration.feature [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        Examples:
+            | testcase | variable                              | value           | field                         |
+            | panic    | BUGSNAG_APP_VERSION                   | 1.4.34          | app.version                   |
+            | panic    | BUGSNAG_APP_TYPE                      | mailer-daemon   | app.type                      |
+            | panic    | BUGSNAG_RELEASE_STAGE                 | beta1           | app.releaseStage              |
+            | panic    | BUGSNAG_HOSTNAME                      | dream-machine-2 | device.hostname               |
+            | panic    | BUGSNAG_METADATA_device_instance      | kube2-33-A      | metaData.device.instance      |
+            | panic    | BUGSNAG_METADATA_framework_version    | v3.1.0          | metaData.framework.version    |
+            | panic    | BUGSNAG_METADATA_device_runtime_level | 1C              | metaData.device.runtime level |
+            | panic    | BUGSNAG_METADATA_Carrot               | orange          | metaData.custom.Carrot        |
```
Comment (imjoehaines) [inline features/configuration.feature:28]: Is setting metadata this way definitely something we want to support? I don't think we do anywhere else

File: v2/configuration.go [hunk 2]
```diff
@@ -262,3 +263,62 @@ func (config *Configuration) notifyInReleaseStage() bool {
...
+			}
+			if keypath, err := parseMetadataKeypath(key); err == nil {
+				addEnvMetadata(event, keypath, value)
+			}
+		}
+
+		return nil
+	})
+
+	config.update(&envConfig)
```
[the remark this answers is not in the document]
Comment (kattrali) [reply v2/configuration.go:323]: This is called from `Configure()` which `update()`s again with the values specified in code, so those take precedence. So in this case:

```go
// env: BUGSNAG_RELEASE_STAGE=beta
bugsnag.Configure(bugsnag.Configuration{ReleaseStage:"alpha"})
```

The release stage would be `alpha`

[the remark this answers is not in the document]
Comment (imjoehaines) [reply v2/configuration.go:323]: Yeah I see, that makes sense

File: v2/configuration.go [hunk 3]
```diff
@@ -262,3 +263,62 @@ func (config *Configuration) notifyInReleaseStage() bool {
...
+
+func (config *Configuration) loadEnv() {
+	envConfig := Configuration{}
+	if apiKey := os.Getenv("BUGSNAG_API_KEY"); apiKey != "" {
+		envConfig.APIKey = apiKey
+	}
+	if endpoint := os.Getenv("BUGSNAG_SESSIONS_ENDPOINT"); endpoint != "" {
+		envConfig.Endpoints.Sessions = endpoint
+	}
+	if endpoint := os.Getenv("BUGSNAG_ENDPOINT"); endpoint != "" {
```
[the remark this answers is not in the document]
Comment (kattrali) [reply v2/configuration.go]: yep!

File: features/configuration.feature (hunk 1 above)
Comment (kattrali) [reply features/configuration.feature]: yeah. no code configuration might be necessary for future changes, though for now I've split this into a separate changeset.

File: v2/configuration.go [hunk 4]
```diff
@@ -262,3 +263,62 @@ func (config *Configuration) notifyInReleaseStage() bool {
...
+	}
+	if autoSessions := os.Getenv("BUGSNAG_AUTO_CAPTURE_SESSIONS"); autoSessions != "" {
+		envConfig.AutoCaptureSessions = autoSessions == "1"
+	}
+	if filters := os.Getenv("BUGSNAG_PARAMS_FILTERS"); filters != "" {
+		envConfig.ParamsFilters = strings.Split(filters, ",")
+	}
+
+	OnBeforeNotify(func(event *Event, config *Configuration) error {
+		for _, value := range os.Environ() {
```
[the remark this answers is not in the document]
Comment (kattrali) [reply v2/configuration.go]: My thought was that someone might add a variable at runtime with `os.Setenv()` but maybe that's overthinking it.

[the remark this answers is not in the document]
Comment (imjoehaines) [reply v2/configuration.go]: Yeah I can see why that might be nice, but I think I'd err on the side of expecting people to add metadata directly in that case

**Review by imjoehaines — APPROVED** [review]

**Pull request merged**